### DISCUSSION TITLE
Can I use GCP Datastream to pull data from Supabase PSQL db?

### DISCUSSION THREAD

**Discussion by Lu1815** [supabase/supabase#29631] [Questions]
There is the need to pull data from the Supabase DB to Bigquery in GCP using Datastream service from GCP. However whenever I try to setup the Datastream I am getting this error:

`Datastream could not create a replication connection to the source database. It appears that Datastream is connected to PgBouncer or Supavisor, but connection pooling is not compatible with logical replication. Edit the source connection profile and provide the IPv4 address and port of the source database instead of the IP and port of the connection pooler, and try again...`

After encountering this error, I checked Supabase to see if there were options to either obtain the IPv4 address or disable connection pooling for the database, but neither of these options is currently available in Supabase.

Having faced this error multiple times today, a question arises: is there an alternative way, or even a method, to use the Supabase PostgreSQL database with Datastream?"

**Comment by GaryAustin1** [comment]
If you are Pro you can for $4 obtain a direct connection with ipv4.

**Reply by Lu1815** [reply]
Do I need to pay a pro suscription to obtain the direct connection with ipv4? Or can I just pay the 4$ for the ipv4 connection?

**Reply by GaryAustin1** [reply]
It requires the Pro plan.

**Comment by cshlin** [comment]
Has anyone gotten this to work? After enabling IPv4, I'm unable to connect from GCP Datastream:

We can't connect to the source database using the specified hostname and port. Make sure that the hostname and port are correct and that the database can accept inbound connections.

I'm able to connect using pooled (which doesn't support logical replication), and using a direct connection from pgAdmin.

**Comment by Rubenburdin** [comment]
Hey Lu1815,
Yeah this is a common friction point when setting up logical replication from Supabase. The error you're seeing is because Datastream needs a direct PostgreSQL connection for CDC, but Supabase routes traffic through Supavisor (their connection pooler) by default. Poolers don't forward the replication protocol, so Datastream can't establish the logical replication slot it needs.
As Gary mentioned, you need the Pro plan plus the IPv4 add-on ($4/month) to get a direct connection endpoint that bypasses the pooler. The direct connection string looks different from the pooled one, make sure you're grabbing it from the Database Settings page under "Direct connection" not "Connection pooling." Supabase docs cover the difference here (https://supabase.com/docs/guides/database/connecting-to-postgres#direct-connection).
@cshlin for your issue connecting even with IPv4 enabled, GCP Datastream uses specific IP ranges that need to be allowed through. You'll want to check your Supabase network restrictions under Project Settings > Database and whitelist the Datastream source IPs for your region. Google publishes those ranges in their Datastream docs. Also double-check you're using port 5432 for direct connection, not 6543 which is the pooler port. If pgAdmin works but Datastream doesn't, it's almost certainly a firewall rule missing the GCP IP ranges.
Hope that helps sort it out
